source/adapters/level_zero/queue.cpp

```diff
@@ -593,11 +603,17 @@ UR_APIEXPORT ur_result_t UR_APICALL urQueueRetain(
   return UR_RESULT_SUCCESS;
 }
 
-UR_APIEXPORT ur_result_t UR_APICALL urQueueRelease(
-    ur_queue_handle_t UrQueue ///< [in] handle of the queue object to release
+UR_APIEXPORT ur_result_t UR_APICALL urQueueRetain(
+    ur_queue_handle_t UrQueue ///< [in] handle of the queue object to get access
 ) {
-  auto Queue = Legacy(UrQueue);
+  return UrQueue->Dispatcher.dispatch(
```

Comment:
I know we are trying to minimize changes and use static dispatch where possible, but looking at this right now, I think simple inheritance from a base abstract class might be much simpler.

Reply:
It might be true, if we want to implement each function for each queue variant then there are not many benefits of using a variant anyway (like template functions), Approach with inheritance would look like this: https://github.com/igchor/unified-runtime/tree/queue_v2_inheritance

Reply:
Hm, both approaches have merit, but I think my preference would be inheritance just because its simpler to grok.

Reply:
Ok, I've just pushed a version with inheritance. The base class and API functions that operate on queue are now auto-generated. The script and templates for generation can be merged or not, but in either case, if we'd like to go back to the variant-approach for some reason, it should be pretty easy to do.